### PR TITLE
Empty state response

### PR DESCRIPTION
**Pull request by Kouprin**
Related to #1052.

### CODE REVIEW

File: chain/chain/src/chain.rs [hunk 1]
```diff
@@ -1141,15 +1141,23 @@ impl Chain {
...
         // 1a. `sync_prev` means the last of the prev epoch.
         // 2. Empty prefix means the height where chunk was applied last time in the prev epoch.
         //    Let's call it `current`.
         // 2a. `prev_` means we're working with height before current.
         // 3. In inner loops we use all prefixes with no relation to the context described above.
-        let sync_block = self.get_block(&sync_hash)?;
+        let sync_block = match self.get_block(&sync_hash) {
+            Ok(sync_block) => sync_block,
+            _ => {
+                // This case may appear in case of latency in epoch switching.
```
Comment (bowenwang1996) [inline chain/chain/src/chain.rs]: Just to make sure that I understands this: are you saying that it is possible that when you sync from a peer they don't have the sync hash block? If that's the case how come we even sync from that peer in the first place? Shouldn't we only sync from peer that has at least newer state than us?

Comment (Kouprin) [reply chain/chain/src/chain.rs]: I guess the following is happening here.
Let's say, current epoch_id = 50.
Node A will care about shard 0 in epoch_id = 51, so it downloads a state now.
Node B will care about shard 0 in epoch_id = 52.
At the moment when epoch switches from 50 to 51, node A will become representative for shard. What's why node B can request a state from node A from the very beginning of the epoch 51. However, there is some latency between blocks transfering, and node A may be at the very end of epoch 50, having no first block for epoch 51.

> Shouldn't we only sync from peer that has at least newer state than us?

I don't mind this is an issue. We just can ask a state in next iteration. It should be recovered quickly.

Comment (bowenwang1996) [reply chain/chain/src/chain.rs]: Well I think it's still better if we can avoid it, no? When we choose which peer to sync from, don't we actually know their head? It should be relatively easy to check whether they are behind.

Comment (Kouprin) [reply chain/chain/src/chain.rs]: Do you know how to do it easily? I guess we should send a separate query to ask "are you ready to sync?" before syncing which absolutely repeats current behavior.

Comment (bowenwang1996) [reply chain/chain/src/chain.rs]: Never mind. I misremembered how state sync works.

File: chain/chain/src/chain.rs [hunk 2]
```diff
@@ -1557,12 +1525,12 @@ impl Chain {
         sync_hash: CryptoHash,
         part_id: u64,
         num_parts: u64,
-        data: &Vec<u8>,
+        data: Vec<u8>,
```
Comment (bowenwang1996) [inline chain/chain/src/chain.rs]: what is this change for?

**Comment by Kouprin** [comment]
> what's the reason for splitting state request into `StateRequestHeader` and `StateRequestPart`?

Code quality. We're having less magic inside.
Technically, those requests are different. One of them is about getting State Header, and the other one is for getting State Part. That's why I decided to make things clearer by splitting original StateRequest.

File: chain/chain/src/chain.rs (hunk 2 above)
Comment (Kouprin) [reply chain/chain/src/chain.rs]: To know it's not borrowing. `set_state_part` indicates it wants to consume the data.

Comment (bowenwang1996) [reply chain/chain/src/chain.rs]: But inside the function `data` is not consumed. Why do you need to pass the value?

Comment (Kouprin) [reply chain/chain/src/chain.rs]: Yep. I guess it should but our storage seems to work different way.
I'm okay with either borrowing or not.

Comment (Kouprin) [reply chain/chain/src/chain.rs]: Ah, I see. That's my fault. I'll revert it.

**Pull request merged**